core/domain/user_id_migration.py

```diff
@@ -103,12 +106,18 @@ def _copy_model_with_new_id_and_user_id(
             model_values['user_id'] = new_user_id
             new_models.append(model_class(**model_values))
 
-        def _replace_models():
+        def _replace_models(new_models_sub, old_models_sub):
             """Replace old models with new ones."""
-            model_class.put_multi(new_models, update_last_updated_time=False)
-            model_class.delete_multi(old_models)
-
-        transaction_services.run_in_transaction(_replace_models)
+            model_class.put_multi(
+                new_models_sub, update_last_updated_time=False)
+            model_class.delete_multi(old_models_sub)
+
+        for i in python_utils.RANGE(
```

Comment:
Add an explanation for why we limit the number of models.

Also, this introduces a possible failure case. What happens if the last model to be migrated fails somehow -- what are our remediation steps in that case?

Also, how many models does this for loop currently operate over?

Reply:
Done.

If anything fails we can always repeat the migration.

I'm not sure, the models that are in this category are for example models that track the progress of the user in an exploration, meaning that there could be around 100 for one user. But to get more precise results we would have to query the datastore.

Reply:
> Also, this introduces a possible failure case. What happens if the last model to be migrated fails somehow -- what are our remediation steps in that case?

Does repeating the migration actually work? If we are partially in a migration of user IDs then in the second run of map(), wouldn't " if user_model.id != user_model.gae_id:" have the possibility of already being True, leading to some orphaned models?

Reply:
Hmmm, true this can happen, but this will be checked by the verification jobs that will be run after the migration. So it is true that the silent errors can happen but will be detected by the verification job and most of the errors we had were visible in the logs.

Reply:
Sure, I understand that. But suppose, when we run this job in production, that the above scenario happens. What should we do then? We should think this through as part of our procedure.

Reply:
The verification will fail, and we restore the backed up data.

Reply:
The verification will fail, and we restore the backed up data.

Reply:
I think that's a problem. Restoring data takes quite a while and this may lead to extended server downtime. Typically what we should be able to do is to just run the job again (i.e. the job should be idempotent). Can you find a way to make it so?

Reply:
@seanlip This should be fixed now. The job is idempotent now.

Reply:
Add individual tests that simulate the different cases for each of the possible user ID migration policies (for dependent models). This is crucial functionality and it is important we get it right and be prepared for any unexpected errors.

E.g. for _copy_model_with_new_id_and_user_id, have tests that:

- Setup a simulation where there are two models of this kind, one has been copied, the other is not (because of an earlier failure). Rerun the job. Check that it doesn't error (even though the first model was deleted after the move), it only migrates one model, etc.
- Setup a simulation where there are two models of this kind, and neither have been copied. Rerun the job. Check that it doesn't error and does the correct thing.
- Setup a simulation where there are two models of this kind, and both have already been copied. Rerun the job. Check that it has no effect on the models and does nothing.

Do similarly comprehensive unit tests for each case. Think through the different breakage possibilities and write one test for each of them. That will give more assurance that we are covered for the different eventualities and that the job is truly idempotent.